README
Wink
====

Wink is a minimalists weblog engine written in Ruby using Rack, Sinatra, and
DataMapper. See the project repository on GitHub for more information:

http://github.com/rtomayko/wink/

Status
------

A complete mess, at the moment. It will likely not run for you at all. The
sources have been made available as-is for discussion purposes only. If you're
mainly interested in running a blog, watch for a 1.0 release. If you're
interested in hacking, stick around...

Requirements
------------

Gems (need all their dependencies of course):

    $ sudo gem install datamapper --version=0.2.5
    $ sudo gem install do_mysql --version=0.2.4
    $ sudo gem install sinatra BlueCloth rubypants html5 test-spec

You may need to download the do_mysql-0.2.4 gem from
http://rubyforge.org/frs/?group_id=4651 and install it manually:
    $ sudo gem install do_mysql-0.2.4.gem

To run the specs you also need SQLite3 and the SQLite3-adapter:
    $ sudo gem install do_sqlite3 --version=0.2.5

We are currently developing on MySQL and SQLite3. Other databases are
theoretically easy to support but we're limiting to these for now to speed
initial development.

Wink is developed primarily under MacOS X and tested on FreeBSD and
a variety of GNU/Linux distributions.

Hacking
-------

Project sources are stored in Git:

    $ git clone git://github.com/rtomayko/wink.git

Create a new MySQL database:

    $ mysql -u root
    mysql> CREATE DATABASE wink_development CHARACTER SET = utf8;

Next, change into your local wink clone and copy the `wink.conf.example`
to `wink.conf`; edit to taste, making sure the database related settings
are adjusted appropriately.

Run the `db:init` task to create the initial table schema and then start a new
development server:

    $ rake db:init
    $ ruby wink

The application should be available at: http://localhost:4567

Testing
-------

Unit and functional tests are kept under the test directory. Before running
tests, you will need to create a separate database:

    $ mysql -u root
    mysql> CREATE DATABASE wink_test CHARACTER SET = utf8;

Once the database has been created, use rake to run the test suite:

    $ rake test

COPYING
-------

All sources included in this distribution are made available under the MIT
license. See the file COPYING for more information.


ChangeLog
== 0.2 / 2008-06-08

* bugfix: Remove rack_cacher
* bugfix: Support for reloading all wink sources in development
* bugfix: Model validators cause errors on reload
* bugfix: Replace "Weblog" OpenStruct with "Wink" module
* bugfix: Unit tests for all models
* bugfix: Comment feed uses hardcoded tag: identifier
* bugfix: Akismet spam detection runs even when no Akismet key is specified
* bugfix: Anonymize CSS and JS sources
* bugfix: Akismet fails with: "undefined method headers"
* Move helpers into module(s) for RDoc
* Make URLs and area names customizable
* Basic testing framework
* Custom Text Filters
* Three phase boot: config, models, web
* Use Sinatra's options for configuration
* Make comment filters customizable ...
* Make log output configurable
* Rename EntryTags model to Tagging
* Use wink.[option] for accessing author (and other options) instead of ivars
* Add sinatra as a submodule
* Remove broken text filters support (for now)
* Better config file organization
* Move Akismet and Delicious helper libs into Wink module

== 0.1 / 2008-05-23
* bugfix: Remove hard-coded Akismet key (cschneid)
* bugfix: Sinatra reloading breaks DM models (cschneid)
* feature: Template Anonymization (cschneid)
* feature: Extract sources from http://tomayko.com
* feature: LICENSE and README


lib/wink.rb
require 'sinatra'
require 'wink/core_extensions'
require 'wink/schema'

# Tell Sinatra to not automatically start a server.
set :run, false

unless reloading?

  # The site's root URL. Note: the trailing slash should be 
  # omitted when setting this option.
  set :url, 'http://localhost:4567'

  # A regular expression that matches URLs to your site's content. Used
  # to detect bookmarks and external content referencing the current site.
  set :url_regex, /http\:/

  # The full name of the site's author.
  set :author, 'Anonymous Coward'

  # The administrator username. You will be prompted to authenticate with
  # this username before modifying entries and comments or providing other
  # administrative activities. Default: "admin".
  set :username, 'admin'

  # The administrator password (see #username). The password is +nil+ by
  # default, disabling administrative access; you must set the password
  # explicitly.
  set :password, nil

  # The site's Akismet key, if spam detection should be performed.
  set :akismet_key, nil

  # The URL of the site as registered with Akismet. Defaults to the
  # +url+ option.
  set :akismet_url, nil

  # Boolean specifying whether Akismet checks should be performed in all
  # environments. Default is to check w/ Akismet only when in production
  # environment.
  set :akismet_always, false

  # A del.icio.us username/password as a two-tuple: ['username', 'password'].
  # When set, del.icio.us bookmark synchronization may be performed by calling
  # Bookmark.synchronize!
  set :delicious, nil

  # Where to write log messages.
  set :log_stream, STDERR

  # The site's title. Defaults to the author's name.
  set :title, nil

  # Title of area that lists Article entries.
  set :writings, 'Writings'

  # Title of area that lists Bookmark entries.
  set :linkings, 'Linkings'

  # Start date for archives + copyright notice.
  set :begin_date, Date.today.year

  # List of filters to apply to comments.
  set :comment_filters, [:markdown, :sanitize]

  # List of patterns that should cause a comment to be marked as spam. The
  # blacklist check occurs before akismet checking.
  set :comment_blacklist, nil

  # URL mappings for various sections of the site
  set :writings_url, "/writings/"
  set :linkings_url, "/linkings/"
  set :archive_url , "/circa/"
  set :tag_url     , "/topics/"
  set :drafts_url  , "/drafts/"
end


module Wink
  extend self

  VERSION = '0.2'

  # An OpenStruct with all application options.
  def options
    Sinatra.application.options
  end

  # Get an option value.
  def [](option_name)
    options.send(option_name)
  end

  # Set an option value.
  def []=(option_name, value)
    options.send("#{option_name}=", value)
  end

  # Respond to any option attributes defined on the underlying #options
  # object and also to all setter messages.
  def respond_to?(name, include_private=false)
    super ||
      options.respond_to?(name, include_private) ||
      name.to_s[-1] == ?=
  end

  # Delegate all messages to the underlying #options object.
  def method_missing(name, *args, &block)
    options.__send__(name, *args, &block)
  end

  private :method_missing


  # Options ====================================================================

  def akismet_url
    self[:url]
  end


  # Configuration ==============================================================

  # Load configuration from the file specified and/or by executing the block. If
  # both a file and block are given, the config file is loaded first and then
  # the block is executed.
  #
  # Database configuration must be in place once the config file and block are
  # processed.
  def configure(file=nil)
    Kernel::load(file) if file
    yield options if block_given?
    require 'wink/models'
    self
  end

  # Load configuration from the file and/or block as specified in Wink#configure
  # and setup Sinatra to start a server instance.
  def run!(config_file=nil, &block)
    configure(config_file, &block)
    require 'wink/web'
    Sinatra.application.options.run = true
  end

  # Rackup compatible constructor. Use in Rackup files as follows:
  #
  #   require 'wink'
  #   run Wink do |config|
  #     config.env = :production
  #     config.url = 'http://example.com'
  #   end
  #
  # If neither a config_file or a block is given, load the default
  # config file: 'wink.conf'.
  def new(config_file=nil, &block)
    config_file ||= 'wink.conf' unless block_given?
    configure(config_file, &block)
    require 'wink/web'
    Sinatra.application
  end

end
